pkg/controller/service/servicecontroller_test.go

```diff
@@ -295,4 +311,408 @@ func TestGetNodeConditionPredicate(t *testing.T) {
 	}
 }
 
-// TODO(a-robinson): Add tests for update/sync/delete.
+func TestUpdateService(t *testing.T) {
+
+	var controller *ServiceController
+	var cloud *fakecloud.FakeCloud
+
+	testCases := []struct {
+		testName   string
+		key        string
+		updateFn   func() //Manupulate the structure
+		srv        *v1.Service
+		expectedFn func(error, time.Duration) error //Error comparision function
+	}{
+		{
+			testName: "If updating a valid service",
+			key:      "validKey",
+			srv:      defaultExternalService(),
+			updateFn: func() {
+
+				controller, cloud, _ = newController()
+				controller.cache.getOrCreate("validKey")
+
+			},
+			expectedFn: func(err error, retryDuration time.Duration) error {
+
+				if err != nil {
+					return err
+				}
+				if retryDuration != doNotRetry {
+					return fmt.Errorf("retryDuration Expected=%v Obtained=%v", doNotRetry, retryDuration)
+				}
+				return nil
+			},
+		},
+	}
+
+	for _, tst := range testCases {
+		tst.updateFn()
+		srvCache := controller.cache.getOrCreate(tst.key)
+
+		obtErr, retryDuration := controller.processServiceUpdate(srvCache, tst.srv, tst.key)
+		if err := tst.expectedFn(obtErr, retryDuration); err != nil {
+			t.Errorf("%v processServiceUpdate() %v", tst.testName, err)
+		}
+	}
+
+}
+
+func TestSyncService(t *testing.T) {
+
+	var controller *ServiceController
+	var cloud *fakecloud.FakeCloud
+
+	testCases := []struct {
+		testName    string
+		key         string
+		updateFn    func() //Function to manipulate the controller element to simulate error
+		expectedErr error  //syncService() only returns error
+	}{
+		{
+			testName: "if an invalid service name is synced",
+			key:      "invalid/key/string",
+			updateFn: func() {
+				controller, cloud, _ = newController()
+
+			},
+			expectedErr: fmt.Errorf("unexpected key format: %q", "invalid/key/string"),
+		},
+		//TODO: see if we can add a test for valid but error throwing service, its difficult right now because synCService() currently runtime.HandleError
+		{
+			testName: "if valid service",
+			key:      "external-balancer",
+			updateFn: func() {
+				controller, cloud, _ = newController()
+				srv := controller.cache.getOrCreate("external-balancer")
+				srv.state = defaultExternalService()
+			},
+			expectedErr: nil,
+		},
+	}
+
+	for _, tst := range testCases {
+		var printError bool
+		tst.updateFn()
+		obtainedErr := controller.syncService(tst.key)
+		if obtainedErr != tst.expectedErr {
+			printError = true
+			if obtainedErr != nil && tst.expectedErr != nil {
+				if obtainedErr.Error() == tst.expectedErr.Error() {
+					//Check if the actual error matches, if yes then test PASSed
+					printError = false
+				}
+			}
+		}
+		if printError {
+			t.Errorf("%v Expected=%v Obtained=%v", tst.testName, tst.expectedErr, obtainedErr)
+		}
+	}
+}
+
+func TestDeleteService(t *testing.T) {
+
+	var controller *ServiceController
+	var cloud *fakecloud.FakeCloud
+
+	testCases := []struct {
+		testName        string
```

Comment:
I have moved the key as a global variable as we are using the same key throughout the test cases.